src/basedtyping/__init__.py

```diff
@@ -591,6 +587,116 @@ def deco(fn: Callable[[], None]) -> Callable[[], None]: ...
         @deco
         def foo(): ...
     """
-    if not isinstance(fn, FunctionType):  # type: ignore[redundant-expr]
+    if not isinstance(fn, FunctionType):
         raise TypeError(f"{fn} is not a FunctionType")
-    return fn  # type: ignore[unreachable]
+    # https://github.com/KotlinIsland/basedmypy/issues/745
+    return cast("FunctionType[P, T]", fn)
+
+
+class ForwardRef(typing.ForwardRef, _root=True):  # type: ignore[call-arg,misc]
```

Comment:
I'm not sure how; but maybe it's possible to make this generic on `__forward_value__`?

Reply:
maybe, but i'm only including this as an implementation detail for the visitor, cal look into improving things later